### PR TITLE
feat: add helm support (#539)

### LINKED ISSUE #539: Add Helm Chart support similar to VSCode OpenShift Toolkit
[kind/Feature]

## Which functionality do you think we should add?

Users should have the option to browse and install Helm Charts to the connected cluster. There should be an UI view to list all the Helm Charts and Install button with the chart information.

## Why is this needed?

As a user, I want to install a Helm Chart directly from the plugin and also should be able to browse other charts available to be installed in the cluster. Using Helm Charts, users can install databases and can link to the component created.

### PR DESCRIPTION
**Pull request by adietish**
fixes #539

### CODE REVIEW

**Comment by datho7561** [comment]
I think it looks pretty good. I think dropping the Versions column in order to show more of the Description makes sense.

**Comment by adietish** [comment]
Current implementation tries to avoid a 2nd dialog on top of an existing dialog:

You get the list of available charts listed in a dialog. The chart that is selected in the list is displayed with it's details in the bottom panel (name, versions, description, etc).

image

In this details panel you have an "Install" button which leads you to the actual install form where you provide the name of the installation, the version etc. Once you provided those you can hit "Install" again to trigger the actual installation:

image

Not sure at 100% if this workflow is best of breed. @datho7561, @sbouchet: thoughts?

**Comment by adietish** [comment]
@sbouchet, @datho7561: Can you please test this? it's afaik feature complete. Code review cannot be done fully yet, I still need to add unit-tests. Thanks!

**Comment by sbouchet** [comment]
i don't have a repo configured (starting from scratch) thus the helm chart dialog can't show any helm charts.
need to manually execute `helm repo add openshift https://charts.openshift.io/`. 
for initial usage, with no repo, this hould be done automatically

**Comment by adietish** [comment]
@sbouchet I disabled editing in the table. I'd love if you could also verify if the layout is now word wrapping instead of using a single line (ex. versions in the details pane). 
I also made the dialog resizable by grabbing the borders and movable by dragging the title bar (isn't great though).

**Comment by adietish** [comment]
@sbouchet, @datho7561: I now also added icons. Please review.

**Comment by sbouchet** [comment]
> @sbouchet I disabled editing in the table. I'd love if you could also verify if the layout is now word wrapping instead of using a single line (ex. versions in the details pane). I also made the dialog resizable by grabbing the borders and movable by dragging the title bar (isn't great though).

that's way better. 
Capture d’écran du 2023-11-02 11-24-04

the dialog grabbing/moving is working, but i don't get feedback ( for grabbing, no cursor changes on borders; for moving, no dialog position update while moving ).

**Comment by sbouchet** [comment]
playing around, i'm getting this bug when trying to install one chart that misses some parameters taht fails, then choose another chart. See screencast 
[Capture vidéo du 2023-11-02 11-25-52.webm](https://github.com/redhat-developer/intellij-openshift-connector/assets/600414/abf0ed30-8306-4a34-9a22-fbed20d28692)

**Comment by sbouchet** [comment]
also get some memory leak messages in console : 
```
2023-11-02 11:27:31,267 [ 903237] SEVERE - #c.i.o.u.ObjectTree - Memory leak detected: 'org.jboss.tools.intellij.openshift.ui.helm.ChartPanels[,0,382,876x303,invalid,alignmentX=0.0,alignmentY=0.0,border=,flags=0,maximumSize=,minimumSize=,preferredSize=]' (class org.jboss.tools.intellij.openshift.ui.helm.ChartPanels) was registered in Disposer as a child of 'ROOT_DISPOSABLE' (class com.intellij.openapi.util.Disposer$2) but wasn't disposed.
Register it with a proper parentDisposable or ensure that it's always disposed by direct Disposer.dispose call.
See https://jetbrains.org/intellij/sdk/docs/basics/disposers.html for more details.
The corresponding Disposer.register() stacktrace is shown as the cause:

java.lang.RuntimeException: Memory leak detected: 'org.jboss.tools.intellij.openshift.ui.helm.ChartPanels[,0,382,876x303,invalid,alignmentX=0.0,alignmentY=0.0,border=,flags=0,maximumSize=,minimumSize=,preferredSize=]' (class org.jboss.tools.intellij.openshift.ui.helm.ChartPanels) was registered in Disposer as a child of 'ROOT_DISPOSABLE' (class com.intellij.openapi.util.Disposer$2) but wasn't disposed.
Register it with a proper parentDisposable or ensure that it's always disposed by direct Disposer.dispose call.
See https://jetbrains.org/intellij/sdk/docs/basics/disposers.html for more details.
The corresponding Disposer.register() stacktrace is shown as the cause:

	at com.intellij.openapi.util.ObjectNode.assertNoChildren(ObjectNode.java:45)
	at com.intellij.openapi.util.ObjectTree.assertIsEmpty(ObjectTree.java:218)
	at com.intellij.openapi.util.Disposer.assertIsEmpty(Disposer.java:278)
	at com.intellij.openapi.util.Disposer.assertIsEmpty(Disposer.java:272)
	at com.intellij.openapi.application.impl.ApplicationImpl.disposeContainer(ApplicationImpl.java:223)
	at com.intellij.openapi.application.impl.ApplicationImpl.lambda$doExit$5(ApplicationImpl.java:660)
	at com.intellij.platform.diagnostic.telemetry.helpers.TraceKt.computeWithSpanIgnoreThrows(trace.kt:73)
	at com.intellij.platform.diagnostic.telemetry.helpers.TraceUtil.computeWithSpanThrows(TraceUtil.java:18)
	at com.intellij.openapi.application.impl.ApplicationImpl.doExit(ApplicationImpl.java:646)
	at com.intellij.openapi.application.impl.ApplicationImpl.exit(ApplicationImpl.java:595)
	at com.intellij.openapi.application.impl.ApplicationImpl.exit(ApplicationImpl.java:584)
	at com.intellij.openapi.application.ex.ApplicationEx.exit(ApplicationEx.java:77)
	at com.intellij.ide.actions.ExitAction.actionPerformed(ExitAction.java:27)
	at com.intellij.openapi.actionSystem.ex.ActionUtil.doPerformActionOrShowPopup(ActionUtil.java:339)
	at com.intellij.openapi.actionSystem.ex.ActionUtil.lambda$performActionDumbAwareWithCallbacks$4(ActionUtil.java:313)
	at com.intellij.openapi.actionSystem.ex.ActionUtil.performDumbAwareWithCallbacks(ActionUtil.java:362)
	at com.intellij.openapi.actionSystem.ex.ActionUtil.performActionDumbAwareWithCallbacks(ActionUtil.java:313)
	at com.intellij.openapi.actionSystem.impl.ActionMenuItem.lambda$performAction$5(ActionMenuItem.java:300)
	at com.intellij.openapi.wm.impl.FocusManagerImpl.runOnOwnContext(FocusManagerImpl.java:226)
	at com.intellij.openapi.actionSystem.impl.ActionMenuItem.performAction(ActionMenuItem.java:293)
	at com.intellij.openapi.actionSystem.impl.ActionMenuItem.lambda$new$0(ActionMenuItem.java:68)
	at java.desktop/javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:1972)
	at com.intellij.openapi.actionSystem.impl.ActionMenuItem.lambda$fireActionPerformed$4(ActionMenuItem.java:115)
	at com.intellij.openapi.application.TransactionGuardImpl.performActivity(TransactionGuardImpl.java:105)
	at com.intellij.openapi.application.TransactionGuardImpl.performUserActivity(TransactionGuardImpl.java:94)
	at com.intellij.openapi.actionSystem.impl.ActionMenuItem.fireActionPerformed(ActionMenuItem.java:115)
	at com.intellij.ui.plaf.beg.BegMenuItemUI.doClick(BegMenuItemUI.java:526)
	at com.intellij.ui.plaf.beg.BegMenuItemUI$MyMouseInputHandler.mouseReleased(BegMenuItemUI.java:558)
	at java.desktop/java.awt.Component.processMouseEvent(Component.java:6657)
	at java.desktop/javax.swing.JComponent.processMouseEvent(JComponent.java:3385)
	at java.desktop/java.awt.Component.processEvent(Component.java:6422)
	at java.desktop/java.awt.Container.processEvent(Container.java:2266)
	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:5027)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2324)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4855)
	at java.desktop/java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4954)
	at java.desktop/java.awt.LightweightDispatcher.processMouseEvent(Container.java:4581)
	at java.desktop/java.awt.LightweightDispatcher.dispatchEvent(Container.java:4522)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2310)
	at java.desktop/java.awt.Window.dispatchEventImpl(Window.java:2808)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4855)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:791)
	at java.desktop/java.awt.EventQueue$3.run(EventQueue.java:740)
	at java.desktop/java.awt.EventQueue$3.run(EventQueue.java:734)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:97)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:764)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:762)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:761)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.kt:685)
	at com.intellij.ide.IdeEventQueue.dispatchMouseEvent(IdeEventQueue.kt:633)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.kt:588)
	at com.intellij.ide.IdeEventQueue.access$_dispatchEvent(IdeEventQueue.kt:67)
	at com.intellij.ide.IdeEventQueue$dispatchEvent$processEventRunnable$1$1$1.compute(IdeEventQueue.kt:369)
	at com.intellij.ide.IdeEventQueue$dispatchEvent$processEventRunnable$1$1$1.compute(IdeEventQueue.kt:368)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:787)
	at com.intellij.ide.IdeEventQueue$dispatchEvent$processEventRunnable$1$1.invoke(IdeEventQueue.kt:368)
	at com.intellij.ide.IdeEventQueue$dispatchEvent$processEventRunnable$1$1.invoke(IdeEventQueue.kt:363)
	at com.intellij.ide.IdeEventQueueKt.performActivity$lambda$1(IdeEventQueue.kt:992)
	at com.intellij.openapi.application.TransactionGuardImpl.performActivity(TransactionGuardImpl.java:113)
	at com.intellij.ide.IdeEventQueueKt.performActivity(IdeEventQueue.kt:992)
	at com.intellij.ide.IdeEventQueue.dispatchEvent$lambda$7(IdeEventQueue.kt:363)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:861)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.kt:405)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:207)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:128)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:117)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:105)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:92)
Caused by: java.lang.Throwable
	at com.intellij.openapi.util.ObjectNode.<init>(ObjectNode.java:24)
	at com.intellij.openapi.util.ObjectNode.findOrCreateChildNode(ObjectNode.java:140)
	at com.intellij.openapi.util.ObjectTree.register(ObjectTree.java:51)
	at com.intellij.openapi.util.Disposer.register(Disposer.java:161)
	at com.intellij.openapi.ui.ComponentValidator.installOn(ComponentValidator.java:155)
	at org.jboss.tools.intellij.openshift.ui.helm.InstallPanel.initComponents(InstallPanel.java:116)
	at org.jboss.tools.intellij.openshift.ui.helm.InstallPanel.<init>(InstallPanel.java:93)
	at org.jboss.tools.intellij.openshift.ui.helm.ChartPanels.create(ChartPanels.java:48)
	at org.jboss.tools.intellij.openshift.ui.helm.ChartPanels.create(ChartPanels.java:25)
	at com.intellij.ui.CardLayoutPanel.createValue(CardLayoutPanel.java:73)
	at com.intellij.ui.CardLayoutPanel.select(CardLayoutPanel.java:101)
	at com.intellij.ui.CardLayoutPanel.lambda$selectLater$0(CardLayoutPanel.java:132)
	at com.intellij.openapi.application.TransactionGuardImpl.runWithWritingAllowed(TransactionGuardImpl.java:208)
	at com.intellij.openapi.application.TransactionGuardImpl.access$100(TransactionGuardImpl.java:21)
	at com.intellij.openapi.application.TransactionGuardImpl$1.run(TransactionGuardImpl.java:190)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:861)
	at com.intellij.openapi.application.impl.ApplicationImpl$4.run(ApplicationImpl.java:478)
	at com.intellij.openapi.application.impl.FlushQueue.doRun(FlushQueue.java:79)
	at com.intellij.openapi.application.impl.FlushQueue.runNextEvent(FlushQueue.java:121)
	at com.intellij.openapi.application.impl.FlushQueue.flushNow(FlushQueue.java:41)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:318)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:789)
	at java.desktop/java.awt.EventQueue$3.run(EventQueue.java:740)
	at java.desktop/java.awt.EventQueue$3.run(EventQueue.java:734)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:759)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.kt:685)
	at com.intellij.ide.IdeEventQueue._dispatchEvent$lambda$10(IdeEventQueue.kt:589)
	at com.intellij.openapi.application.impl.ApplicationImpl.runWithoutImplicitRead(ApplicationImpl.java:1485)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.kt:589)
	at com.intellij.ide.IdeEventQueue.access$_dispatchEvent(IdeEventQueue.kt:67)
	at com.intellij.ide.IdeEventQueue$dispatchEvent$processEventRunnable$1$1$1.compute(IdeEventQueue.kt:369)
	at com.intellij.ide.IdeEventQueue$dispatchEvent$processEventRunnable$1$1$1.compute(IdeEventQueue.kt:368)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:787)
	at com.intellij.ide.IdeEventQueue$dispatchEvent$processEventRunnable$1$1.invoke(IdeEventQueue.kt:368)
	at com.intellij.ide.IdeEventQueue$dispatchEvent$processEventRunnable$1$1.invoke(IdeEventQueue.kt:363)
	at com.intellij.ide.IdeEventQueueKt.performActivity$lambda$1(IdeEventQueue.kt:992)
	at com.intellij.openapi.application.TransactionGuardImpl.performActivity(TransactionGuardImpl.java:105)
	... 10 more
```

**Comment by adietish** [comment]
> 
> that's way better. 

awesome, glad this worked, finally.

> the dialog grabbing/moving is working, but i don't get feedback ( for grabbing, no cursor changes on borders; for moving, no dialog position update while moving ).

Ok, great. There's only feedback for resizing, grabbing doesn't. Can't do anything about it. The Jetbrains "Find UI" doesn't provide feedback either, but I noticed how their's is way smoother.

**Comment by adietish** [comment]
> playing around, i'm getting this bug when trying to install one chart that misses some parameters taht fails, then choose another chart. 

To my understanding this is a bug/problem in the repo. The chart is listed but is not installable, imho there isn't anything I can do on my side to fix this, it's a problem of the helm binary/helm repo.

**Comment by sbouchet** [comment]
> > playing around, i'm getting this bug when trying to install one chart that misses some parameters taht fails, then choose another chart.
> 
> To my understanding this is a bug/problem in the repo. The chart is listed but is not installable, imho there isn't anything I can do on my side to fix this, it's a problem of the helm binary/helm repo.

please see my steps from the screencast. i'm trying to install the vert.x chart that does not requires any params, and get an error about installing the **wildfly** chart that i previously selected. 

if i try to install direclty the vertx chart, that works as expected.

**Comment by sbouchet** [comment]
@adietish : last version works great ! the tree refresh on uninstall is smooth :)

File: src/main/.DS_Store
Comment (sbouchet) [inline src/main/.DS_Store:1]: please ignore those files

File: src/main/java/org/jboss/tools/intellij/openshift/actions/HelmAction.java
Comment (sbouchet) [inline src/main/java/org/jboss/tools/intellij/openshift/actions/HelmAction.java:2]: should be 2023

File: src/main/java/org/jboss/tools/intellij/openshift/tree/application/ApplicationsRootNode.java
Comment (sbouchet) [inline src/main/java/org/jboss/tools/intellij/openshift/tree/application/ApplicationsRootNode.java:82]: remove semicolon

**Comment by adietish** [comment]
It would be nice to be able to navigate the table that contains the charts by pressind cursor up & down. I'll file an enhancement issue.

File: src/main/java/org/jboss/tools/intellij/openshift/tree/application/ApplicationsTreeStructure.java
Comment (sbouchet) [inline src/main/java/org/jboss/tools/intellij/openshift/tree/application/ApplicationsTreeStructure.java:231]: shouldn't be a null check for odo variable after that call ?

File: src/main/java/org/jboss/tools/intellij/openshift/ui/SwingUtils.java
Comment (sbouchet) [inline src/main/java/org/jboss/tools/intellij/openshift/ui/SwingUtils.java:2]: should be 2023

File: src/main/java/org/jboss/tools/intellij/openshift/utils/ToolFactory.java
Comment (sbouchet) [inline src/main/java/org/jboss/tools/intellij/openshift/utils/ToolFactory.java:2]: should be 2023

**Review by sbouchet — COMMENTED** [review]
Overral LGTM, small changes needed

**Comment by adietish** [comment]
@sbouchet I'm looking into the failing IT. Shouldn't we fix it before merging?

**Comment by sbouchet** [comment]
> @sbouchet I'm looking into the failing IT. Shouldn't we fix it before merging?

no, it's part of the **UI** integration Test, that are new, added by #535 , and failling since.

**Comment by adietish** [comment]
> no, it's part of the **UI** integration Test, that are new, added by #535 , and failling since.
@sbouchet ok, please re-approve and I merged it 🎉

**Review by sbouchet — APPROVED** [review]
LGTM, good work on this !

**Comment by adietish** [comment]
> LGTM, good work on this !

thanks :) appreciate :)

**Comment by adietish** [comment]
@sbouchet still can't merge it, what am I missing?

**Pull request merged**
